org.eclipse.xtext.util/src/org/eclipse/xtext/util/UriExtensions.java

```diff
@@ -8,35 +8,103 @@
  */
 package org.eclipse.xtext.util;
 
-import java.net.URISyntaxException;
+import java.util.Arrays;
 
 import org.eclipse.emf.common.util.URI;
 
+import com.google.inject.Singleton;
+
 /**
  * @author Christian Dietrich - Initial contribution and API
  * @since 2.14
  */
+@Singleton
 public class UriExtensions {
 	/**
-	 * Returns an URI with empty authority, if authority is absent (null) and has file scheme.
+	 * Converts the given absolute, encoded stringUri to an EMF URI. If the URI
+	 * is a {@link URI#isFile() file URI}, it is guaranteed that its
+	 * {@link URI#authority() authority} is not null.
 	 */
 	public URI toUri(String stringUri) {
-		try {
-			java.net.URI netUri = new java.net.URI(stringUri);
-			return toEmfUri(netUri);
-		} catch (URISyntaxException e) {
-			throw new RuntimeException(e);
+		URI result = URI.createURI(stringUri);
+		// Short-circuit the good case where the URI doesn't contain any unusual escape sequences
+		if (stringUri.contains("%")) {
+			return cleanupEncodedCharacters(result);
+		}
+		return withEmptyAuthority(result);
+	}
+
+	private URI cleanupEncodedCharacters(URI encoded) {
+		// Decompose the URI
+		String authority = encoded.authority();
+		String[] segments = encoded.segments();
+		String device = encoded.device();
+		String opaque = encoded.opaquePart();
+		String query = encoded.query();
+		String fragment = encoded.fragment();
```

Comment:
do these include all parts of an uri like userinfo?

Reply:
I was under the impression, yes. I'll double-check.

Reply:
User-Info is part of the authority. Should be fine.